src/rgb/schema/data.rs

```diff
@@ -510,3 +527,254 @@ mod strict_encoding {
+
+                (
+                    Self::PublicKey(EllipticCurve::Secp256k1, _),
+                    data::Revealed::Secp256k1Pubkey(_),
+                ) => {}
+                (
+                    Self::PublicKey(EllipticCurve::Curve25519, _),
+                    data::Revealed::Ed25519Pubkey(_),
+                ) => {}
+                (
+                    Self::Signature(elliptic_curve::SignatureAlgorithm::Ecdsa),
+                    data::Revealed::Secp256k1ECDSASignature(_),
+                ) => {}
+                (
+                    Self::Signature(elliptic_curve::SignatureAlgorithm::Ed25519),
+                    data::Revealed::Ed25519Signature(_),
+                ) => {}
+
+                _ => {
+                    status.add_failure(validation::Failure::SchemaMismatchedDataType(item_id));
+                }
+            }
+            status
+        }
+    }
+
+    impl StateFormat {
+        pub fn validate<STATE>(
+            &self,
+            node_id: &NodeId,
+            assignment_id: usize,
+            data: &Assignment<STATE>,
+        ) -> validation::Status
+        where
+            STATE: StateTypes,
+            EncodingError: From<<STATE::Confidential as StrictEncode>::Error>
+                + From<<STATE::Confidential as StrictDecode>::Error>
+                + From<<STATE::Revealed as StrictEncode>::Error>
+                + From<<STATE::Revealed as StrictDecode>::Error>,
+        {
+            let mut status = validation::Status::new();
+            match data {
+                Assignment::Confidential { assigned_state, .. }
+                | Assignment::ConfidentialAmount { assigned_state, .. } => {
+                    let a: &dyn Any = assigned_state.as_any();
+                    match self {
+                        StateFormat::Void => {
+                            if a.downcast_ref::<<VoidStrategy as StateTypes>::Confidential>()
+                                .is_none()
+                            {
+                                status.add_failure(validation::Failure::SchemaMismatchedStateType(
+                                    assignment_id,
+                                ));
+                            }
+                        }
+                        StateFormat::Homomorphic(_) => {
+                            if a.downcast_ref::<<HomomorphStrategy as StateTypes>::Confidential>()
+                                .is_none()
+                            {
+                                status.add_failure(validation::Failure::SchemaMismatchedStateType(
+                                    assignment_id,
+                                ));
+                            }
+                            // TODO: When other homomorphic formats will be added,
+                            //       add information to the status like with
+                            //       hashed data below
+                        }
```

Comment:
Are you actually validating the content of these or just checking the type? I couldn't find the part where confidential amounts are summed together and checked.

I looked for references to `verify_commit_sum()` but I couldn't find any, are you using a different function?

Reply:
This will be a part of Virtual machine "predefined function". It will be called from here: https://github.com/LNP-BP/rust-lnpbp/blob/9fe45094aee8ff361fb37f1e51f642f7afc48b6f/src/rgb/schema/schema.rs#L251